bot/core/paginator.py

```diff
@@ -0,0 +1,421 @@
+
+import discord
+from discord import Embed
+from discord.abc import Messageable
+from discord.ext import commands
+from discord.ext.commands import Context, Paginator
+
+logger = logging.getLogger(__name__)
+
+
+class PaginatorError(Exception):
+    """
+    Exception Class for Custom error PaginatorError.
+    Raised if the bot is unable to create/process a paginator.
+    """
+    pass
+
+
+def check_permissions(channel: Messageable, bot: commands.Bot) -> bool:
+    """
+    Checks if the bot has required permissions to send or process
+    the paginator
+
+    Args:
+        channel (Messageable): The Context
+        bot (commands.Bot): The bot object, used for using bot.user if there is no guild
+
+    Returns:
+        bool: Whether the bot has all of the required permissions
+    """
+    perms: discord.Permissions = channel.permissions_for(
+        channel.guild.me if channel.guild is not None else bot.user
+    )
+
+    necessary_permissions = [
+        perms.send_messages,
+        perms.embed_links,
+        perms.add_reactions,
+        perms.read_message_history
+    ]
+
+    return all(necessary_permissions)
+
+
+class _BasePaginator:
+    """
+    An embed can be added to the embeds list
+    if the paginator class is inheriting from
+    this class
+    """
+    def __init__(self, embeds: t.List[Embed]) -> None:
+        self.embeds = embeds
+
+    def __add__(self, other: Embed) -> None:
+        self.embeds.append(other)
+
+
+class FieldPaginator(_BasePaginator):
+    """
+    A class which creates a list of embeds and configures their
+    fields according to the parameters.
+
+    Parameters
+    ----------
+        *args : t.Union[str, Embed]
+            Brief Explanation: The arguments can be a list,
+            representing a page or it can be an embed.
+
+            Example:
+            Key-Value is required. Inline is optional since its
+            set to True as a default value.
+            To insert an embed (creating a new page),
+            create a new value on the list as shown below.
+
+            FieldPaginator([{'Page': '1', 'inline': True}, discord.Embed(title="Page 2")])
+    """
+    def __init__(self, *args) -> None:
+        self.embeds = []
+        super().__init__(self.embeds)
+
+        for index, arg in enumerate(args):
+            if isinstance(arg, Embed):
+                self.embeds.append(arg)
+            elif isinstance(arg, list):
+                _embed = discord.Embed()
+                for element in arg:
+                    if isinstance(element, dict):
+                        # If element is an empty dictionary
+                        if element == {}:
+                            raise PaginatorError(
+                                "FieldPaginator argument is a nested dictionary but does not have the mandatory "
+                                "key-value pair".strip()
+                            )
+                        elif len(element.items()) > 2:
```

Comment:
There is no need for `elif` here, we can use `if` since the purpose of `elif` is to only run if the first `if` won't pass, in this case, we don't need to check for that since if it will pass the function will raise an exception and therefore avoiding the need to check if the first if was run.